From 6d0b312a9a18db9767cb2f48523a128e32eab2e2 Mon Sep 17 00:00:00 2001
From: "Langer, Thomas" <tlanger@maxlinear.com>
Date: Mon, 19 Apr 2021 19:10:53 +0200
Subject: [PATCH] PONRTSYS-8555: devm-cleanup in dp_mib_counters.c

Do not use devm functions to allocate temporary buffers that are  freed
in the same function again.
---
 .../net/datapath/mib_counters/dp_mib_counters.c    | 23 +++++++++-------------
 1 file changed, 9 insertions(+), 14 deletions(-)

diff --git a/drivers/net/datapath/mib_counters/dp_mib_counters.c b/drivers/net/datapath/mib_counters/dp_mib_counters.c
index a189ab80774c..1d7399528a3f 100644
--- a/drivers/net/datapath/mib_counters/dp_mib_counters.c
+++ b/drivers/net/datapath/mib_counters/dp_mib_counters.c
@@ -220,9 +220,7 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 	if (!cnt)
 		return;
 
-	tmp_counters = devm_kzalloc(&g_dp_dev->dev,
-				    (2 * sizeof(GSW_RMON_Port_cnt_t)),
-				    GFP_ATOMIC);
+	tmp_counters = kzalloc(2 * sizeof(GSW_RMON_Port_cnt_t), GFP_ATOMIC);
 	if (!tmp_counters)
 		return;
 
@@ -251,7 +249,7 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 		ret = ops->gsw_rmon_ops.RMON_Port_Get(ops, &tmp_counters[i]);
 		if (ret != DP_SUCCESS) {
 			pr_err("MIB CNT ERROR - RMON_Port_Get - %d\n", ret);
-			devm_kfree(&g_dp_dev->dev, tmp_counters);
+			kfree(tmp_counters);
 			return;
 		}
 	}
@@ -297,8 +295,7 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 	/* mac_ops is optional, do not fail if unavailable */
 	mac_ops = gsw_get_mac_ops(cnt->inst, cnt->port_id);
 	if (mac_ops) {
-		rmon = devm_kzalloc(&g_dp_dev->dev, sizeof(struct mac_rmon),
-				    GFP_ATOMIC);
+		rmon = kzalloc(sizeof(*rmon), GFP_ATOMIC);
 		if (rmon) {
 			mac_ops->rmon_get(mac_ops, rmon);
 			tmp_counters[0].nRxGoodPausePkts =
@@ -374,9 +371,8 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 	write_unlock(&cnt->cnt_lock);
 
 free_exit:
-	devm_kfree(&g_dp_dev->dev, tmp_counters);
-	if (rmon)
-		devm_kfree(&g_dp_dev->dev, rmon);
+	kfree(tmp_counters);
+	kfree(rmon);
 }
 
 /* check if is it the one that is search for */
@@ -713,8 +709,7 @@ void mib_cnt_get_ethtool_stats_31(struct net_device *dev,
 	u32 lct_idx = 0;
 	int ret;
 
-	ps_subif = devm_kzalloc(&g_dp_dev->dev, (sizeof(dp_subif_t)),
-				GFP_ATOMIC);
+	ps_subif = kzalloc(sizeof(dp_subif_t), GFP_ATOMIC);
 	if (!ps_subif)
 		return;
 
@@ -722,7 +717,7 @@ void mib_cnt_get_ethtool_stats_31(struct net_device *dev,
 	ret = dp_get_netif_subifid(dev, NULL, NULL, NULL, ps_subif, flags);
 	if (ret != DP_SUCCESS) {
 		netdev_err(dev, "dp_get_netif_subifid returned %d\n", ret);
-		devm_kfree(&g_dp_dev->dev, ps_subif);
+		kfree(ps_subif);
 		return;
 	}
 
@@ -737,12 +732,12 @@ void mib_cnt_get_ethtool_stats_31(struct net_device *dev,
 	mib_cnt = find_registered_dev(ps_subif->port_id, sub_if_id_grp,
 				      ps_subif->inst, dev);
 	if (!mib_cnt) {
-		devm_kfree(&g_dp_dev->dev, ps_subif);
+		kfree(ps_subif);
 		read_unlock(&dev_list_lock);
 		return;
 	}
 
-	devm_kfree(&g_dp_dev->dev, ps_subif);
+	kfree(ps_subif);
 
 	/* Update counters for the requested net device */
 	update_dev_counters(mib_cnt, !INITIAL_UPDATE);
